ipalib/util.py

```diff
@@ -187,6 +198,137 @@ def normalize_zone(zone):
+
+    if max_version_idx < min_allowed_idx:
+        max_version_idx = min_version_idx
+        root_logger.warning("tls_version_max set too low ('{old}'),"
+                            "using '{new}' instead"
+                            .format(old=tls_version_max,
+                                    new=TLS_VERSIONS[max_version_idx]))
+    return TLS_VERSIONS[min_version_idx:max_version_idx+1]
+
+
+def create_https_connection(
+    host, port=HTTPSConnection.default_port,
+    cafile=None,
+    client_certfile=None, client_keyfile=None,
+    keyfile_passwd=None,
+    tls_version_min="tls1.1",
+    tls_version_max="tls1.2",
+    **kwargs
+):
+    """
+    Create a customized HTTPSConnection object.
+
+    :param host:  The host to connect to
+    :param port:  The port to connect to, defaults to
+               HTTPSConnection.default_port
+    :param cafile:  A PEM-format file containning the trusted
+                    CA certificates
+    :param client_certfile:
+            A PEM-format client certificate file that will be used to
+            identificate the user to the server.
+    :param client_keyfile:
+            A file with the client private key. If this argument is not
+            supplied, the key will be sought in client_certfile.
+    :param keyfile_passwd:
+            A path to the file which stores the password that is used to
+            encrypt client_keyfile. Leave default value if the keyfile
+            is not encrypted.
+    :returns An established HTTPS connection to host:port
+    """
+    # pylint: disable=no-member
+    tls_cutoff_map = {
+        "ssl2": ssl.OP_NO_SSLv2,
+        "ssl3": ssl.OP_NO_SSLv3,
+        "tls1.0": ssl.OP_NO_TLSv1,
+        "tls1.1": ssl.OP_NO_TLSv1_1,
+        "tls1.2": ssl.OP_NO_TLSv1_2,
+    }
+    # pylint: enable=no-member
+
+    if cafile is None:
+        raise RuntimeError("cafile argument is required to perform server "
+                           "certificate verification")
+
+    # remove the slice of negating protocol options according to options
+    tls_span = get_proper_tls_version_span(tls_version_min, tls_version_max)
+
+    # official Python documentation states that the best option to get
+    # TLSv1 and later is to setup SSLContext with PROTOCOL_SSLv23
+    # and then negate the insecure SSLv2 and SSLv3
+    # pylint: disable=no-member
+    ctx = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
+    ctx.options |= (
+        ssl.OP_ALL | ssl.OP_NO_COMPRESSION | ssl.OP_SINGLE_DH_USE |
+        ssl.OP_SINGLE_ECDH_USE | ssl.OP_NO_SSLv2 | ssl.OP_NO_SSLv3
+    )
+
+    # pylint: enable=no-member
+    # disable all TLS versions not in tls_span
+    for version in TLS_VERSIONS:
+        if version not in tls_span:
+            ctx.options |= tls_cutoff_map[version]
+
+    # make sure the required TLS versions are available if Python decides
+    # to modify the default TLS flags
+    for version in tls_span:
+        ctx.options &= ~tls_cutoff_map[version]
```

Comment:
Nitpick: it looks like this can be done in a single `for` loop rather than two:
```python
    for version in TLS_VERSIONS:
        if version in tls_spawn:
            ctx.options &= ~tls_cutoff_map[version]
        else:
            ctx.options |= ~tls_cutoff_map[version]
```